### PR TITLE
Support network in initrd and entering LUKS passphrase via SSH

### PR DESCRIPTION
**Pull request by abbradar**
This pile of changes adds three main things to NixOS:

1) Support network in initrd; possible use cases may be SSH (implemented here) or network root filesystem:
2) Support SSH in initrd via the new network support;
3) Support entering LUKS passphrase via SSH.

This calls for a flurry of changes:

1) Add new `mkinitcpio-nfs-utils` package from Arch, which provides a nice little utility for parsing `ip=` kernel parameter and setting up a connection;
2) Add a patch for `dropbear` to propagate `LD_LIBRARY_PATH` (needed for running things via SSH);
3) Add support for excluding paths from nuking via `nuke-references`; this adds a `perl` dependency to it;
4) Use (3) to remove workarounds from `stage-1` for when `$out` is nuked too;
6) Add a new hook for extra commands in `stage-1` named `postEarlyDeviceCommands`, called before `preLVMCommands` (where passphrase request occurs); it's used to initialize networking and start SSH server.
6) Add a new NixOS module which makes necessary changes to `initrd` to support networking and SSH; we may want some things moved into core `initrd` later, such as `mount devpts`;
7) Finally, add a wrapper around `cryptsetup` to `luksroot` which saves its arguments and allows one to call `cryptsetup` from another place simultaneously; when one succeeds, others (waiting for a passphrase which was already entered) are killed.

This allows one to encrypt root partition on a remote server and enter the passphrase via SSH, inspired by `dropbear_initrd_encrypt` AUR package in Arch Linux. When `cryptsetup` asks for a passphrase, you may log in via SSH, run `cryptsetup-askpass`, enter the phrase and startup will proceed automatically.

The main problem I see is that SSH private keys are stored unprotected in the Nix store, so this can be somewhat dangerous. Possible attack scenario is:

1) Gain access to read arbitrary files on the filesystem (with any user), steal SSH private keys;
2) Wait until server is rebooted; Intercept SSH connection with a proxy and record the passphrase;
3) Physically remove the drive (or get access to the raw partition via another means) and decrypt it.

This can be prevented if the keys are stored somewhere where regular users can't access them; any ideas how can we accomplish this?

cc @nckx (`dropbear` changes), @edolstra (`nuke-references` changes); I'm not sure who to ping w.r.t. `luksroot` changes.

### CODE REVIEW

**Comment by peti** [comment]
Very nice! This is a feature I wanted to have for a long time!

**Comment by abbradar** [comment]
I've thought more about this; in my attack description, the attacker needs to be able to extract raw disk image, so even if we hide SSH keys from regular users, the attack can still be performed with extra (0) step "Extract raw disk image; extract SSH keys from unencrypted initrd which must be somewhere on the disk". So I guess this risk is inseparable from the feature, but requirements for simultaneous disk image and network channel access makes this difficult enough.

**Comment by geerds** [comment]
Great work @abbradar ! :beer:

**Comment by nckx** [comment]
This is a wonderful feature. More beverage of your choice to you!

\**glances wistfully at https://github.com/NixOS/nix/issues/8**

**Comment by geerds** [comment]
This would also fix #4196

**Comment by abbradar** [comment]
Have someone tested this already? It would be good to have people test this, since this touches critical things (i.e. initrd).

**Comment by abbradar** [comment]
Apart from one possibly bad design decision (add of a new initrd hook) this shouldn't break anything and I've been using this for some time; there's people interested so I guess I'll wait for a couple of days and then merge it.

**Comment by abbradar** [comment]
Let's hope it wouldn't break anything...

**Comment by edolstra** [comment]
Hm, doesn't this make a machine have different host keys during stage 1 and stage 2? That seems undesirable...

**Comment by abbradar** [comment]
Yes it does, but we can't do anything about it without seriously compromising security of main SSH keys (by storing them in the Nix store _and_ unencrypted `/boot`).

**Comment by abbradar** [comment]
One idea is to use different non-standard port for SSH in stage-1. This way if I remember correctly it would count as a different "host" in OpenSSH's `known_hosts` sense, allowing for a different key.

**Comment by CMCDragonkai** [comment]
Awesome, I was looking for this. Have you seen this variant: http://blog.hintcafe.com/post/85987577654/encrypting-linux-root-disk-in-ec2cloud it allows a more automated way of getting keys too.

**Comment by CMCDragonkai** [comment]
By the way what about WLAN support at the initrd?

**Comment by abbradar** [comment]
@CMCDragonkai No, I haven't seen this -- I'll take a look later. Thanks!

WLAN support would be tricky and would require additional custom utilities, I don't have time to try for now.

File: nixos/modules/system/boot/initrd-network.nix [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      type = types.int;
+      default = 22;
+      description = ''
+        Port on which SSH initrd service should listen.
+      '';
+    };
+
+    boot.initrd.network.ssh.shell = mkOption {
+      type = types.str;
+      default = "/bin/ash";
```
Comment (msteen) [inline nixos/modules/system/boot/initrd-network.nix]: Should this not be "/bin/bash"?

**Comment by nilsmeyer** [comment]
I've been trying to use this with an encrypted root partition, however since the SSH Keys are linked to are not accessible. I think it would be better to copy the dropbear SSH host keys to the initrd.

**Comment by abbradar** [comment]
Very strange. How do you specify a key file?

**Comment by nilsmeyer** [comment]
For lack of a better idea where to put them:
`boot.initrd.network.ssh.hostRSAKey = "/etc/ssh/dropbear_host_rsa_key";`
`boot.initrd.network.ssh.hostECDSAKey = "/etc/ssh/dropbear_host_ecdsa_key";`

In the initrd they aren't there of course since /etc/ssh isn't available yet because the root fs isn't mounted. I'm currently trying to figure out how to copy files to the initrd with a deterministic path.

**Comment by abbradar** [comment]
Oh, that's because you need to specify such files directly so that they would be copied to Nix store:

`boot.initrd.network.ssh.hostRSAKey = /etc/ssh/dropbear_host_rsa_key;`

Yes, I understand this makes server' private keys world-readable. This feature should be used with understanding and care (for example, always use different SSH keys for this than your main ones!) before NixOS/nix#8 is implemented...

**Comment by nilsmeyer** [comment]
Oh I see, thank you.

**Pull request merged**